### PR TITLE
Base for Utility Objects

### PR DESCRIPTION
**Pull request by SteelFall**
The classes added and changes in pre-existing classes done in this PR should provide a good base for future Utility objects. Most of the new classes are copied from their furniture counterparts and will need cleaning up once we are sure that certain features are not needed in those classes.

### CODE REVIEW

File: Assets/Scripts/Controllers/BuildModeController.cs [hunk 1]
```diff
@@ -222,6 +230,72 @@ public void DoBuild(Tile t)
...
+                        Tile offsetTile = WorldController.Instance.world.GetTileAt(x_off,y_off);
+                        offsetTile.PendingBuildJob = j;
+                        j.cbJobStopped += (theJob) =>
+                            {
+                                offsetTile.PendingBuildJob = null;
+                            };
+                    }
+                }
+
+                // Add the job to the queue
```
Comment (Randshot) [inline Assets/Scripts/Controllers/BuildModeController.cs]: Comments should end with a period.
This also applies to following lines:
- 235
- 244
- 253
- 258
- 277

File: Assets/Scripts/Models/Tile.cs [hunk 2]
```diff
@@ -151,6 +158,56 @@ public bool PlaceFurniture(Furniture objInstance)
         return true;
     }
 
+    public bool UnplaceUtility()
+    {
+        // Just uninstalling.
+
+        if (Utility == null)
+            return false;
```
Comment (Randshot) [inline Assets/Scripts/Models/Tile.cs]: Maybe add the curly braces anyway?

File: Assets/Scripts/Models/Tile.cs [hunk 3]
```diff
@@ -151,6 +158,56 @@ public bool PlaceFurniture(Furniture objInstance)
...
+
+        Utility u = Utility;
+
+        for (int x_off = X; x_off < (X + u.Width); x_off++)
+        {
+            for (int y_off = Y; y_off < (Y + u.Height); y_off++)
+            {
+
```
Comment (Randshot) [inline Assets/Scripts/Models/Tile.cs]: I think that this is an unnecessary empty line.

File: Assets/Scripts/Models/Tile.cs [hunk 4]
```diff
@@ -151,6 +158,56 @@ public bool PlaceFurniture(Furniture objInstance)
...
+        }
+        
+        for (int x_off = X; x_off < (X + objInstance.Width); x_off++)
+        {
+            for (int y_off = Y; y_off < (Y + objInstance.Height); y_off++)
+            {
+
+                Tile t = World.current.GetTileAt(x_off, y_off);
+                t.Utility = objInstance;
+
```
Comment (Randshot) [inline Assets/Scripts/Models/Tile.cs]: Also here, why all these empty lines?

File: Assets/Scripts/Models/Utility.cs [hunk 5]
```diff
@@ -0,0 +1,874 @@
...
+
+    /// <summary>
+    /// These actions are called every update. They get passed the utility
+    /// they belong to, plus a deltaTime.
+    /// </summary>
+    // protected Action<Utility, float> updateActions;
+    protected List<string> updateActions;
+
+    /// <summary>
+    /// These context menu lua action are used to build the context menu of the utility
```
Comment (Randshot) [inline Assets/Scripts/Models/Utility.cs]: Period.

File: Assets/Scripts/Models/Utility.cs [hunk 6]
```diff
@@ -0,0 +1,874 @@
...
+        {
+            return PowerValue < 0.0f;
+        }
+    }
+
+    public void Update(float deltaTime)
+    {
+        if (updateActions != null)
+        {
+            // updateActions(this, deltaTime);
```
Comment (Randshot) [inline Assets/Scripts/Models/Utility.cs]: Avoid pushing commented out code.

File: Assets/Scripts/Models/Utility.cs [hunk 7]
```diff
@@ -0,0 +1,874 @@
...
+    }
+
+    public ENTERABILITY IsEnterable()
+    {
+        if (isEnterableAction == null || isEnterableAction.Length == 0)
+        {
+            return ENTERABILITY.Yes;
+        }
+
+        //// UtilityActions.CallFunctionsWithUtility( isEnterableActions.ToArray(), this );
```
Comment (Randshot) [inline Assets/Scripts/Models/Utility.cs]: Avoid pushing commented out code.

**Comment by Randshot** [comment]
@SteelFall After rebasing your branch to the latest commit on the TeamPorcupine/master branch, I got one conflict. The conflict itself is easily resolvable, but when I tried to run it in Unity3D I got this error:
```
Assets/Scripts/Controllers/BuildModeController.cs(289,46): error CS1061: Type `WorldController' does not contain a definition for `devMode' and no extension method `devMode' of type `WorldController' could be found (are you missing a using directive or an assembly reference?)
```

**Comment by SteelFall** [comment]
@Randshot Going to try to fix things now.

**Pull request closed without merging**